zkevm-circuits/src/evm_circuit/util/common_gadget.rs

```diff
@@ -1039,3 +1039,125 @@ impl<F: Field> CommonErrorGadget<F> {
         Ok(1u64)
     }
 }
+
+/// Check if the passed in word is within the specified byte range and less than
+/// a maximum cap.
+#[derive(Clone, Debug)]
+pub(crate) struct WordByteCapGadget<F, const VALID_BYTES: usize> {
+    word: WordByteRangeGadget<F, VALID_BYTES>,
+    lt_cap: LtGadget<F, VALID_BYTES>,
+}
+
+impl<F: Field, const VALID_BYTES: usize> WordByteCapGadget<F, VALID_BYTES> {
+    pub(crate) fn construct(cb: &mut ConstraintBuilder<F>, cap: Expression<F>) -> Self {
+        let word = WordByteRangeGadget::construct(cb);
+        let value = select::expr(word.within_range(), word.valid_value(), cap.expr());
+        let lt_cap = LtGadget::construct(cb, value, cap);
+
+        Self { word, lt_cap }
+    }
+
+    /// Return true if within the specified byte range, false if overflow. No
+    /// matter whether it is less than the cap.
+    pub(crate) fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        original: U256,
+        cap: F,
+    ) -> Result<bool, Error> {
+        let within_range = self.word.assign(region, offset, original)?;
+
+        let value = if within_range {
+            let mut bytes = [0; 32];
+            bytes[0..VALID_BYTES].copy_from_slice(&original.to_le_bytes()[0..VALID_BYTES]);
+            F::from_repr(bytes).unwrap()
+        } else {
+            cap
+        };
+
+        self.lt_cap.assign(region, offset, value, cap)?;
+
+        Ok(within_range)
+    }
+
+    pub(crate) fn lt_cap(&self) -> Expression<F> {
+        self.lt_cap.expr()
+    }
+
+    pub(crate) fn original_word(&self) -> Expression<F> {
+        self.word.original_word()
+    }
+
+    pub(crate) fn overflow(&self) -> Expression<F> {
+        self.word.overflow()
+    }
+
+    pub(crate) fn valid_value(&self) -> Expression<F> {
+        self.word.valid_value()
+    }
+
+    pub(crate) fn within_range(&self) -> Expression<F> {
+        self.word.within_range()
+    }
+}
+
+/// Check if the passed in word is within the specified byte range.
+#[derive(Clone, Debug)]
+pub(crate) struct WordByteRangeGadget<F, const VALID_BYTES: usize> {
+    original: Word<F>,
+    within_range: IsZeroGadget<F>,
+}
+
+impl<F: Field, const VALID_BYTES: usize> WordByteRangeGadget<F, VALID_BYTES> {
+    pub(crate) fn construct(cb: &mut ConstraintBuilder<F>) -> Self {
+        debug_assert!(VALID_BYTES < 32);
+
+        let original = cb.query_word_rlc();
+        let within_range = IsZeroGadget::construct(cb, sum::expr(&original.cells[VALID_BYTES..]));
+
+        Self {
+            original,
+            within_range,
+        }
+    }
+
+    /// Return true if within the range, false if overflow.
+    pub(crate) fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        original: U256,
+    ) -> Result<bool, Error> {
+        debug_assert!(VALID_BYTES < 32);
```

Comment:
I think this `debug_assert` and the one in `valid_value` will never be triggered: this type has a single constructor `construct` which already has this assert, so if we manage to create the the variable it means `VALID_BYTES < 32`.  So any time we call `foo.assign()` or `foo.valid_value()` we know for sure that `VALID_BYTES < 32`, because it has been checked at `let foo = WordByteRangeGadget::construct()`.

Reply:
Delete `debug_assert!(VALID_BYTES < 32)` in `assign` and `valid_value` functions (only leave it in `construct`) in commit [29cf64c](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1317/commits/29cf64cc21e305a9248a649f1f8e7fb89917f985).